prov/shm/src/smr_atomic.c

```diff
@@ -357,19 +380,19 @@ static ssize_t smr_atomic_inject(struct fid_ep *ep_fid, const void *buf,
 	rma_ioc.count = count;
 	rma_ioc.key = key;
 
-	smr_generic_format(cmd, peer_id, ofi_op_atomic, 0, 0, 0);
-	smr_generic_atomic_format(cmd, datatype, op);
-
 	if (total_len <= SMR_MSG_DATA_LEN) {
-		smr_format_inline_atomic(cmd, FI_HMEM_SYSTEM, 0, &iov, 1, NULL, 0);
+		smr_do_atomic_inline(ep, peer_smr, id, peer_id, ofi_op_atomic,
+			 	     0, FI_HMEM_SYSTEM, 0, datatype, op,
+				     &iov, 1, total_len);
 	} else if (total_len <= SMR_INJECT_SIZE) {
-		tx_buf = smr_freestack_pop(smr_inject_pool(peer_smr));
-		smr_format_inject_atomic(cmd, FI_HMEM_SYSTEM, 0, &iov, 1, NULL,
-					 0, NULL, 0, peer_smr, tx_buf);
+		ret = smr_do_atomic_inject(ep, peer_smr, id, peer_id,
+				ofi_op_atomic, 0, FI_HMEM_SYSTEM, 0, datatype,
+				op, &iov, 1, NULL, 0, NULL, 0, total_len,
+				NULL, 0);
+		if (ret)
+			goto unlock_region;
```

Comment:
what happens if total_len > SMR_INJECT_SIZE?  There is 'assert(count <= SMR_INJECT_SIZE)' at the beginning of this function, but count is different from total_len.

Reply:
Ah that assert is definitely incorrect. Thanks for catching!